### PR TITLE
Fix integer overflows in attention & cache ops

### LINKED ISSUE #1486: Array Index Overflow for Large #Blocks
Suppose we have 64K blocks, and the key cache shape of a single layer is `[64K, 32, 16, 16, 8]` (LLaMA-7B). For slot mapping, the maximum value is `64K * 32 * 16 * 16 * 8 = 4194304K > INT_MAX`, which overflow.

## Env
`PyTorch 2.0.1 + vLLM 0.2.0 + cuda-11.7`

## Simple reproduce
We can reproduce this by changing the parameters of `tests/kernels/test_cache.py` to the following
```
DTYPES = [torch.half]
NUM_TOKENS = [7]  # Arbitrary values for testing
NUM_LAYERS = [1]  # Arbitrary values for testing
NUM_HEADS = [32]  # Arbitrary values for testing
HEAD_SIZES = [128]
BLOCK_SIZES = [16]
NUM_BLOCKS = [64000]  # Arbitrary values for testing
NUM_MAPPINGS = [32]  # Arbitrary values for testing
SEEDS = [0]
```
The `reshape_and_cache` kernel cannot pass the test.
If we change the dtype of `slot_mapping` to int64, we can pass the test.

Besides, change `NUM_BLOCKS` to 64K in `tests/kernels/test_attention.py`, I find it cannot pass the test, either.
It reports "illegal memory access".

@zhuohan123 @WoosukKwon Can you help look into this?

### PR DESCRIPTION
**Pull request by WoosukKwon**
Fixes #1486 

This PR fixes the overflows in paged attention & cache ops when the number of blocks is huge.

### CODE REVIEW

File: csrc/cache_kernels.cu [hunk 1]
```diff
@@ -141,46 +141,46 @@ namespace vllm {
...
-                            + head_idx * (head_size / x) * block_size * x
-                            + x_idx * block_size * x
-                            + block_offset * x
-                            + x_offset;
-    const int tgt_value_idx = block_idx * num_heads * head_size * block_size
-                              + head_idx * head_size * block_size
-                              + head_offset * block_size
-                              + block_offset;
+    const int64_t tgt_key_idx = block_idx * num_heads * (head_size / x) * block_size * x
+                                + head_idx * (head_size / x) * block_size * x
```
Comment (zhuohan123) [inline csrc/cache_kernels.cu:176]: All the numbers in this expression is `int`: `head_idx * (head_size / x) * block_size * x`. For safety, should we change all numbers to `int64_t`? I think this should only bring negligible overhead but I'm not 100% sure.

**Review by zhuohan123 — APPROVED** [review]
LGTM! Left a small comment about one potential overflow point.

File: csrc/cache_kernels.cu (hunk 1 above)
Comment (WoosukKwon) [reply csrc/cache_kernels.cu:176]: @zhuohan123 Does this relieve your concern?

**Comment by JF-D** [comment]
Thanks for the quick fix!!!

**Pull request merged**